slf4j-api/src/main/java/org/slf4j/TransferableMdc.java

```diff
@@ -0,0 +1,176 @@
+ * <p>
+ * <i>Incorrect 2</i>:
+ * <pre>{@code
+ * TransferableMdc mdc = TransferableMdc.current();
+ * executor.submit(() -> {//task1
+ *  try (TransferableMdc mdcTmp = mdc.apply()) {
+ *      //...
+ *  }
+ * });
+ * executor.submit(() -> {//task2
+ *  try (TransferableMdc mdcTmp = mdc.apply()) {
+ *      //...
+ *  }
+ * });
+ * }</pre>
+ * This is incorrect because {@code task1} and {@code task2} may be executed concurrently and hence
+ * methods {@link #apply()} and {@link #close()} may be executed concurrently, but these methods are not thread-safe.
+ * The simple way to fix the incorrect example is as follows:
+ * <pre>{@code
+ * TransferableMdc mdc1 = TransferableMdc.current();
+ * executor.submit(() -> {//task1
+ *  try (TransferableMdc mdcTmp = mdc1.apply()) {
+ *      //...
+ *  }
+ * });
+ * TransferableMdc mdc2 = TransferableMdc.current();
+ * executor.submit(() -> {//task2
+ *  try (TransferableMdc mdcTmp = mdc2.apply()) {
+ *      //...
+ *  }
+ * });
+ * }</pre>
+ */
+//@NotThreadSafe
+public final class TransferableMdc implements MDCAdapter, Closeable {
+    /**
+     * Creates {@link TransferableMdc} which holds current {@linkplain Thread thread}'s
+     * {@link MDC} {@linkplain MDC#getCopyOfContextMap() context map}.
+     *
+     * @return New {@link TransferableMdc} which state is identical to the current
+     * {@linkplain Thread thread}'s state of {@link MDC}.
+     */
+    public static final TransferableMdc current() {
+        return new TransferableMdc(MDC.getCopyOfContextMap());
+    }
+
+    //@Nullable
+    private final Map<String, String> context;
+    //@Nullable
+    private Map<String, String> backup;
+
+    private TransferableMdc(/*@Nullable*/ final Map<String, String> contextMap) {
+        context = (contextMap == null || contextMap.isEmpty()) ? null : new HashMap<String, String>(contextMap);
+    }
+
+    /**
+     * Merges this {@link TransferableMdc} into the current {@link Thread}'s {@link MDC}
+     * and retains original state of the current {@link MDC}.
+     *
+     * @return {@code this}.
+     */
+    public final TransferableMdc apply() {
+        if (context != null) {
+            //@Nullable
+            final Map<String, String> currentContext = MDC.getCopyOfContextMap();
+            backup = (currentContext == null || currentContext.isEmpty()) ? null : currentContext;
+            for (final Map.Entry<String, String> contextEntry : context.entrySet()) {
+                MDC.put(contextEntry.getKey(), contextEntry.getValue());
+            }
+        }
+        return this;
+    }
+
+    /**
+     * Restores current {@link Thread}'s {@link MDC} to its state before {@link #apply()}
+     * (this state was retained by {@link #apply()}).
+     */
+    //@Override
+    public void close() {
+        if (context != null) {
+            MDC.clear();
+            if (backup != null) {
+                MDC.setContextMap(backup);
+                backup = null;
+            }
+        }
```

Comment:
Thanks @efdknittlfrank ! I fixed this and added `TransferableMdcTest.restoreEmpty` that tests the case you described.

Besides:
- added a not guaranteed check of multiple calls to `TransferableMdc.close`. Such calls signify that `TransferableMdc` is not used correctly.
- renamed `apply` to `transfer` (looks more intuitive), and changed the behaviour so that instead of merging contexts by overriding values of conflicting keys, it resets the context to the transferred state. I can't remember why I originally decided to merge contexts instead. Does anyone have ideas on what behaviour is preferable: merge vs reset? It's also possible to support both and let the user decide, I am just not sure how useful it would be.

Reply:
In my opinion, restoring the MDC to its old state does not makes sense. 

In the case of servlet filters which can be cascaded, it makes sense for the latest filter not to clobber its predecessors when starting. Similarly, when exiting it should only clear keys that it has set itself, leaving other keys alone. However, if it has set a given key, I don't see a use case where it makes sense to restore that key to its old value.

Reply:
@ceki I am not sure I understand the situation you described. Specifically, I don't what is specific about asynchronous handling of requests in filters/servlets comparing to asynchronous handling of other tasks:

- The majority of projects use synchronous processing in `Filter.doFilter` and `Servlet.service`, and in such a situations `TransrefableMdc` is not required (there is no other thread a user may want to transfer the current MDC to). For the situations when `Filter.doFilter`/`Servlet.service` handle requests asynchronously by starting async processing via `ServletRequest.startAsync`, the `FilterChain.doFilter` is still called sequentially. So also no place for `TransrefableMdc`.

- As far a I understand, the only situation when `TransrefableMdc` can be used with filters or servlets, is when after calling `ServletRequest.startAsync` a user submits an async task either via `AsyncContext.start` or by calling `ExecutorService.submit` on a user-controlled executor service (or something similar), but in this case it's not different from submitting any task to any executor.

If you still see something special about servlet filters, could you please describe the situation in more details (maybe a code snippet?).

The idea of `TransrefableMdc.close` is simple: execute a task and leave thread's state (MDC) exactly as it was before executing the task. Why? - So that other tasks don't observe any changes made not by them. Imagine a general-purpose executor (e.g. `ForkJoinPool.commonPool`, but it can be anything). I may want to transfer MDC when submitting a task there. Someone else (someone else's code) may want to just execute the task knowing that MDC is supposed to be empty because he never messed with it in executor's threads. However since `TransrefableMdc` does not rollback MDC changes (what you are proposing?), the second user may now observe unexpected changes made to MDC previously by my task.

Reply:
The way I see the behavior of `TransferableMdc` is that it creates a "Stack" of MDCs. Each `apply` (now `transferAnd…`) operation, pushes to the stack, and each `close` pops the stack, restoring the previous value.
In my eyes, that's sensible behavior (I'm currently relying on it)

Reply:
While we can interpret `TransferableMdc ` as a stack where you can push (`transfer`) not more than one element, and you can pop (`close`) not more than one element. I want to propose a different view of it. But before that, let's consider the situation you probably had in mind when imagining a stack: multiple levels of "nesting".

**Nesting**
For multiple levels of asynchronous nesting we can write the following terrible code:
```
TransferableMdc outermostMdc = TransferableMdc.current();
executor1.submit(() -> {
  try (TransferableMdc transferredMdc = outermostMdc.transfer()) {
    TransferableMdc outerMdc = TransferableMdc.current();
    executor2.submit(() -> {
        try (TransferableMdc weNeedToGoDeeperMdc = outerMdc.transfer()) {
            //is this deep enough?
        }
    }
  }
});
```
I agree, that multiple levels of asynchronous nesting may happen in practice, but the code above is not how it should be handled. A better code would look like the following (the example is artificial, but I believe shows the point):
```
void process(Request req) {
  //...
  submitMdcAware(executor1, () -> processFirstPart(req));
}

void processFirstPart(Request req) {
  //...
  submitMdcAware(executor2, () -> processSecondPart(req));
}

void processSecondPart(req) {
  //...
}

void submitMdcAware(ExecutorService executor, Runnable action) {
  TransferableMdc outerMdc = TransferableMdc.current();
  executor.submit(() -> {
    try (TransferableMdc transferredMdc = outerMdc.transfer()) {
      action.run();
    }
  });
}
```
So the code may be using `TransferableMdc` without even knowing it (see `process` and `processFirstPart`), and multiple levels of asynchrony are handled trivially without supporting any stack functionality in the `TransferableMdc`.

**A (correct? :) view on `TransferableMdc` which is different from seeing it as a stack**
There is no problem in transferring MDC between threads. In fact, both Java's inner classes and lambda expressions capture effectively final variables from their context, thus making the transfer of MDC between threads automatic (as soon as you have a copy of it's map and you access it from a lambda expression). What we are missing is the ability to apply the transferred context (that's why I called the method `apply` originally) and roll it back once it's not needed anymore. And this is what `TransferableMcd` does for a user in a most compact way I can see possible in Java (if someone sees what the library can do to make to code shorter without using proxies or code generation, please share).

P.S. Speaking of proxies, it's definitely possible to for example create `MdcAwareExecutorService implements ExecutorService` (it's method `submit` would look similar to `submitMdcAware` above), so that a user would not need to do anything and MDC would be transferred, applied and rolled back transparently. But SL4J is not a right place for such a class, while `TransferableMcd` provides a building block naturally belonging to SL4J that can be used by others e.g. for building such MDC-aware wrappers.

Reply:
The reason I mentioned servlet filters was to illustrate a case of component composition where each component sets part of the MDC key space. In such a configuration, either the "first" component calls MDC.clear() either at the very beginning or at the very end of processing. In the more general case where the "first" filter cannot be identified, an alternative strategy is for each component to clear the keys it set. In both approaches the idea is to avoid leaving stale data in the thread which might pollute the view of subsequent callers. 

**In light of this concern (i.e. not to leave stale data behind), the idea of restoring previously existing (and most probably stale data) seems quite counterproductive.**

Reply:
@ceki I believe I get what you are saying. Does the following example captures your idea?
```
TransferableMdc outerMdc = TransferableMdc.current();
  executor.submit(() -> {
    /*inner MDC already has the "numberOfSubmittedTasks" key
      initially set to 0 by a thread factory*/
    try (TransferableMdc transferredMdc = outerMdc.transferAndMerge()) {
      MDC.put("numberOfSubmittedTasks",//increment the counter
          Integer.parseInt(MDC.get("numberOfSubmittedTasks")) + 1);
    }
    /*here we still want want to see the incremented value of "numberOfSubmittedTasks"
      so that the next submitted task could also increment it's value,
      instead of always incrementing 0 as it would happen
      if TransferableMdc.close() were rolling back all MDC changes*/
});
}
```

**Merge/unmerge**
I agree that if `TransferableMdc` merges the transferred MDC with the inner MDC via `transferAndMerge`, the behaviour of `close` should be as you mentioned: remove new keys introduced by merging, leave all other keys (and their potentially updated values) that were in the inner MDC before merging (essentially, it's an unmerge operation). Do you agree with this?

**Override/rollback**
The following question still needs to be answered: what should the behaviour of `close` be when `transferAndOverride` is used? It appears to me that in this case rollback (reinstate the original context) is the correct behaviour. Thoughts?

**A request of example from @ceki**
You also mentioned "_a case of component composition where each component sets part of the MDC key space_". If you still see this as something possible with multiple threads involved, could you please give a specific example (not necessary filter/servlet, just anything artificial)? I am asking because I don't think something like this is even possible in theory (or I don't understand your idea; a specific example will help in either case).

Reply:
My concern is quite unsophisticated. I updated my earlier comment to shorten the discussion.

Reply:
@ceki Thanks for the example. Let's see what we have there.

**The third task**
> // this message should not bear data for "client" but probably will

This is true because MDC is thread-scoped, and the only thing that can be done in order to exclude the possibility of observing the `"client"` key is explicitly removing it from MDC (e.g. by calling `MDC.clear`either at the end of the first task or the beginning of the third task). This situation has nothing to do with `TransferableMdc` and the quoted statement is true regardless of the presence of the second task (the one using `TransferableMdc`).

**The second task**
> // this message should not bear MDC data for "client"

If we want this behaviour, we should use `transferAndOverride` instead of `transferAndMerge`. That's it.

To sum up, the problem in the third task has nothing to do with `TransferableMdc`, and the problem in the second task is solved with `TransferableMdc`. But interestingly, none of the problems is about the behaviour of `TransferableMdc.close`.

Reply:
@ceki Do we still have unresolved concerns about either the behaviour of `close` or the proposed merge/unmerge, override/rollback functionality? If not, I will start implementing that. If yes, then lets continue figuring those out.

Reply:
To leave my 2¢: I would very much depend on the "rollback" (stacklike) behavior of close. After leaving a try-with-resources block, I expect to contain the MDC the same keys and values as it did before the block, regardless of any updates that might have happened inside of the block. (Very much like the `putCloseable` on MDC, but not being limited to a single item and supporting multiple key-value pairs at a time)

Reply:
The problem TransferableMdc.close() does not help with the stale data problem. Moreover, it is not hard to imagine situations where it makes it worse. As I mentioned previously, I don't see the point in restoring potentially stale data. Apparently it is useful for nesting MDCs but why would anyone want nested MDCs? Anyone care to elaborate?

Reply:
Our use case is as follows: Often we have several MDC values which should only be set for a single log entry. All log messages are logged to Elasticsearch, in which each MDC value is a separate property of the document. This allows easy filtering on or visualization of certain log properties.
In such a case, we do not want to lose the original value of the MDC key (which potentially is non-empty). After writing the log, the MDC should be restored to the original values, not cleared fully (because some values should really apply to _all_ log messages). Nesting MDCs simplifies this pattern, without having to manually remember the old values (Thinking about it, storing a copy and restoring it manually could also work in this case?).
Maybe we are abusing the MDC here, but it is a super simple and effective way to enable our logs to be sliced and diced (and visualized). There's StructuredArguments, but they have some shortcomings too.

Reply:
> why would anyone want nested MDCs? Anyone care to elaborate?
For us the situation was as follows:
- receive a new HTTP request
- discover client's address from the request and add it to MDC so that we can see it in log messages related to handling the request
- pass the request into a different thread pool for processing (IO threads do not precess requests)
`TransferableMdc` is used to not loose client's address related to the request as a result of handling the request in a thread different from the one received the request and set the address into MDC.

Can the aforementioned be done without `TransferableMdc`? Sure it can, but we can also do anything we do with `MDC` without having `MDC` in SLF4J. However having `MDC` makes things easier. The same is true for `TransferableMdc`.

> The current implementation of TransferableMdc.close() does not help with the stale data problem.
It's not suppose to help with this problem. This problem exists in `MDC` regardless of the presence of `TransferableMdc`, and can be solved e.g. by calling the `clear` method.

> Moreover, it is not hard to imagine situations where it makes it worse.
It seems like it's hard to imagine: I can't see such a situation and you did not present a plausible example. If you don't agree with my analysis of the example you provided, please point out what specifically you disagree about.

I am claiming that if `close` leaves the state of `MDC` exactly the same as it was before `TransferableMdc` was applied, then it's not possible for external code (the code executed outside of the block where `TransferableMdc` is applier) to observe the fact that `TransferableMdc` is used.

Reply:
The need to transferring MDC from one thread to another is **not** being contested. The argument is about the appropriateness of TransferableMdc.close where it **restores** previous data.  

There is no issue explicitly supporting nested MDC by an API which makes the nested structure explicit. The current proposal sneaks in the nested behavior. I am sorry to say that just bad API design.

Reply:
> that just bad API design and not something I can get behind

I can't disagree or agree with a qualitative statement unless it is accompanied by arguments or I can see the arguments on my own. In this case, I can't see the arguments on my own, so I keep asking you to give an example where the introduction of `TransferableMdc` and specifically the call to `TransferableMdc.close` (which restores the original MDC state) makes the possible behaviours worse than they were before the introduction of `TransferableMdc`.

> it is not hard to imagine situations where it makes it worse: a developer clears the MDC only to have it restored by TransferableMdc.close()

May I see a specific example that you have in mind when describing this situation? Just communicate this idea as code so that I don't have to think out (and potentially make mistakes) of what you think of. And once we have a specific argument expressed in a strict form (code), we will be able to discuss it in a productive manner.